snap/info.go

```diff
@@ -1546,3 +1548,85 @@ func (a AppInfoBySnapApp) Less(i, j int) bool {
 	}
 	return iName < jName
 }
+
+// SnapdAssertionMaxFormatsFromSnapFile returns the supported assertion max
+// formats for the snapd code carried by the given snap, plus its snapd
+// version. This is only applicable to snapd/core or UC20+ kernel snaps.
+// For kernel snaps that are not UC20+ or that do not carry the necessary
+// explicit information yes, this can return nil and "" respectively for
+// maxFormats and snapdVersion.
+func SnapdAssertionMaxFormatsFromSnapFile(snapf Container) (maxFormats map[string]int, snapdVersion string, err error) {
+	info, err := ReadInfoFromSnapFile(snapf, nil)
+	if err != nil {
+		return nil, "", err
+	}
+	var infoFile string
+	missingOK := false
+	typ := info.Type()
+	switch typ {
+	case TypeOS, TypeSnapd:
+		infoFile = "/usr/lib/snapd/info"
+	case TypeKernel:
+		infoFile = "/snapd-info"
+		// some old kernel file will not contain this
+		missingOK = true
+	default:
+		return nil, "", fmt.Errorf("cannot extract assertion max formats information, snaps of type %s do not carry snapd", typ)
+	}
+	b, err := snapf.ReadFile(infoFile)
+	if err != nil {
+		if missingOK && os.IsNotExist(err) {
+			return nil, "", nil
+		}
+		return nil, "", err
+	}
+	ver, flags, err := snapdtool.ParseInfoFile(bytes.NewBuffer(b), fmt.Sprintf("from %s snap", typ))
+	if err != nil {
+		return nil, "", err
+	}
```

Comment:
Could this be moved to a helper function called `readAndParseSnapdInfoFile` or similar?

And then replace that with
```
        ver, flags, err := readAndParseSnapdInfoFile(info)
	if err != nil {
		return nil, "", err
	}
```

Reply:
It needs to take also snapf. I'm -0 mostly because it gets further away from the doc comment that explains some details about it, we could give it one but then it's a bit just repetition as it is not used twice.

I'm interested on the 2nd reviewer opinion on this.

Reply:
Also we need to deal with the kernel case again somehow, so the code after its return will be a bit strange unless we introduce a marker error

Reply:
One suggestion that Philip made was to extract a helper to get the snapdInfoPath location. Something like:
```go
diff --git a/snap/info.go b/snap/info.go
index aabac709b0..c8a1fc3c5c 100644
--- a/snap/info.go
+++ b/snap/info.go
@@ -1549,6 +1549,17 @@ func (a AppInfoBySnapApp) Less(i, j int) bool {
        return iName < jName
 }
 
+func snapdInfoFilePath(typ Type) (string, error) {
+       switch typ {
+       case TypeOS, TypeSnapd:
+               return "/usr/lib/snapd/info", nil
+       case TypeKernel:
+               return "/snapd-info", nil
+       default:
+               return "", fmt.Errorf("cannot extract assertion max formats information, snaps of type %s do not carry snapd", typ)
+       }
+}
+
 // SnapdAssertionMaxFormatsFromSnapFile returns the supported assertion max
 // formats for the snapd code carried by the given snap, plus its snapd
 // version. This is only applicable to snapd/core or UC20+ kernel snaps.
@@ -1560,22 +1571,14 @@ func SnapdAssertionMaxFormatsFromSnapFile(snapf Container) (maxFormats map[strin
        if err != nil {
                return nil, "", err
        }
-       var infoFile string
-       missingOK := false
        typ := info.Type()
-       switch typ {
-       case TypeOS, TypeSnapd:
-               infoFile = "/usr/lib/snapd/info"
-       case TypeKernel:
-               infoFile = "/snapd-info"
-               // some old kernel file will not contain this
-               missingOK = true
-       default:
-               return nil, "", fmt.Errorf("cannot extract assertion max formats information, snaps of type %s do not carry snapd", typ)
+       infoFile, err := snapdInfoFilePath(typ)
+       if err != nil {
+               return nil, "", err
        }
        b, err := snapf.ReadFile(infoFile)
        if err != nil {
-               if missingOK && os.IsNotExist(err) {
+               if typ == TypeKernel && os.IsNotExist(err) {
                        return nil, "", nil
                }
                return nil, "", err
```
which does feel like a nice win to me :)